File: apps/mobile/android/app/src/main/kotlin/io/github/obliviousodin/fabric/mobile/core/WorkSync.kt
package io.github.obliviousodin.fabric.mobile.core

/** Scope of one physical gateway/profile Work ledger. */
data class WorkSyncScope(
    val gatewayId: String,
    val profileId: String,
)

enum class WorkProjectionPhase { EMPTY, BOOTSTRAPPING, SYNCING, CURRENT }

data class WorkUnknownProjectionSubject(
    val subjectId: String,
    val subjectType: String,
    val version: Long,
    val subject: WorkEventSubject,
)

/**
 * Serializable projection of the authoritative Work ledger.
 *
 * The `(gatewayId, profileId, ledgerId, cursor)` tuple is state authority.
 * `work.changed` transport hints must cause a sync request, never mutate this
 * object directly.
 */
data class WorkProjection(
    val gatewayId: String,
    val profileId: String,
    val ledgerId: String? = null,
    val cursor: Long? = null,
    val watermark: Long? = null,
    val phase: WorkProjectionPhase = WorkProjectionPhase.EMPTY,
    val nextPageToken: String? = null,
    val resetLedgerHint: String? = null,
    val jobs: Map<String, WorkJobSummary> = emptyMap(),
    val attention: Map<String, WorkAttention> = emptyMap(),
    val unknownSubjects: Map<String, WorkUnknownProjectionSubject> = emptyMap(),
    /** Includes tombstone versions, preventing stale events from resurrecting a row. */
    val subjectVersions: Map<String, Long> = emptyMap(),
)

enum class WorkSyncApplyErrorCode {
    IDENTITY_CHANGED,
    BOOTSTRAP_SEQUENCE_INVALID,
    BOOTSTRAP_REQUIRED,
    CURSOR_INVALID,
    LEDGER_CHANGED,
    PAGE_NON_ACTIONABLE,
}

class WorkSyncApplyException(
    val code: WorkSyncApplyErrorCode,
    message: String,
) : IllegalStateException(message)

data class WorkSyncRequestContext(
    val gatewayId: String,
    val profileId: String,
    /** Token submitted for this bootstrap page; null only for page one. */
    val pageToken: String? = null,
    /** Cursor submitted for this delta page. */
    val after: Long? = null,
)

private fun nonemptyWorkScope(value: String, field: String): String {
    if (value.trim().isEmpty()) throw IllegalArgumentException("$field must be non-empty.")
    return value
}

private fun checkedScope(scope: WorkSyncScope): WorkSyncScope = WorkSyncScope(
    gatewayId = nonemptyWorkScope(scope.gatewayId, "gateway_id"),
    profileId = nonemptyWorkScope(scope.profileId, "profile_id"),
)

private fun checkedContext(context: WorkSyncRequestContext): WorkSyncScope = checkedScope(
    WorkSyncScope(context.gatewayId, context.profileId),
)

fun createWorkProjection(scope: WorkSyncScope): WorkProjection {
    val checked = checkedScope(scope)
    return WorkProjection(gatewayId = checked.gatewayId, profileId = checked.profileId)
}

private fun sameScope(state: WorkProjection, scope: WorkSyncScope): Boolean =
    state.gatewayId == scope.gatewayId && state.profileId == scope.profileId

private fun subjectKey(type: String, id: String): String = "$type:$id"

private data class MutableProjectionSubjects(
    val jobs: MutableMap<String, WorkJobSummary>,
    val attention: MutableMap<String, WorkAttention>,
    val unknown: MutableMap<String, WorkUnknownProjectionSubject>,
    val versions: MutableMap<String, Long>,
)

private fun mutableSubjects(state: WorkProjection? = null): MutableProjectionSubjects =
    MutableProjectionSubjects(
        jobs = state?.jobs?.toMutableMap() ?: mutableMapOf(),
        attention = state?.attention?.toMutableMap() ?: mutableMapOf(),
        unknown = state?.unknownSubjects?.toMutableMap() ?: mutableMapOf(),
        versions = state?.subjectVersions?.toMutableMap() ?: mutableMapOf(),
    )

private fun applyJob(subjects: MutableProjectionSubjects, job: WorkJobSummary) {
    val key = subjectKey("job", job.jobId)
    if ((subjects.versions[key] ?: 0) >= job.version) return
    subjects.jobs[job.jobId] = job
    subjects.unknown.remove(key)
    subjects.versions[key] = job.version
}

private fun applyAttention(subjects: MutableProjectionSubjects, attention: WorkAttention) {
    val key = subjectKey("attention", attention.attentionId)
    if ((subjects.versions[key] ?: 0) >= attention.version) return
    subjects.attention[attention.attentionId] = attention
    subjects.unknown.remove(key)
    subjects.versions[key] = attention.version
}

private fun applyEvent(subjects: MutableProjectionSubjects, event: WorkEvent) {
    val key = subjectKey(event.subjectType, event.subjectId)
    if ((subjects.versions[key] ?: 0) >= event.subjectVersion) return

    if (event.tombstone) {
        when (event.subjectType) {
            "job" -> subjects.jobs.remove(event.subjectId)
            "attention" -> subjects.attention.remove(event.subjectId)
        }
        subjects.unknown.remove(key)
        subjects.versions[key] = event.subjectVersion
        return
    }

    val subject = event.subject ?: throw WorkSyncApplyException(
        WorkSyncApplyErrorCode.PAGE_NON_ACTIONABLE,
        "A non-tombstone work event has no subject.",
    )
    when (subject) {
        is WorkJobSummary -> applyJob(subjects, subject)
        is WorkAttention -> applyAttention(subjects, subject)
        is WorkUnknownSubject -> {
            subjects.unknown[key] = WorkUnknownProjectionSubject(
                subjectId = event.subjectId,
                subjectType = event.subjectType,
                version = event.subjectVersion,
                subject = subject,
            )
            subjects.versions[key] = event.subjectVersion
        }
    }
}

private fun finishProjection(
    base: WorkProjection,
    subjects: MutableProjectionSubjects,
): WorkProjection = base.copy(
    jobs = subjects.jobs.toMap(),
    attention = subjects.attention.toMap(),
    unknownSubjects = subjects.unknown.toMap(),
    subjectVersions = subjects.versions.toMap(),
)

private fun applyBootstrap(
    state: WorkProjection,
    page: WorkSyncPage,
    context: WorkSyncRequestContext,
): WorkProjection {
    val requestedToken = context.pageToken
    val firstPage = requestedToken == null
    val subjects = if (firstPage) {
        // A page-one bootstrap is the only flow allowed to replace state identity.
        mutableSubjects()
    } else {
        if (!sameScope(state, WorkSyncScope(context.gatewayId, context.profileId))) {
            throw WorkSyncApplyException(
                WorkSyncApplyErrorCode.IDENTITY_CHANGED,
                "A bootstrap continuation belongs to a different gateway or profile.",
            )
        }
        if (state.phase != WorkProjectionPhase.BOOTSTRAPPING || state.nextPageToken != requestedToken) {
            throw WorkSyncApplyException(
                WorkSyncApplyErrorCode.BOOTSTRAP_SEQUENCE_INVALID,
                "Bootstrap page token does not match the pending page.",
            )
        }
        if (state.ledgerId != page.ledgerId) {
            throw WorkSyncApplyException(
                WorkSyncApplyErrorCode.LEDGER_CHANGED,
                "The Work ledger changed during bootstrap; restart at page one.",
            )
        }
        if (state.watermark != page.watermark) {
            throw WorkSyncApplyException(
                WorkSyncApplyErrorCode.CURSOR_INVALID,
                "The fixed bootstrap watermark changed between pages.",
            )
        }
        mutableSubjects(state)
    }

    // All work happens on local copies. A malformed row cannot publish a
    // cursor or identity before its entire page has applied.
    page.jobs.forEach { applyJob(subjects, it) }
    page.attention.forEach { applyAttention(subjects, it) }
    return finishProjection(
        WorkProjection(
            gatewayId = context.gatewayId,
            profileId = context.profileId,
            ledgerId = page.ledgerId,
            cursor = if (page.hasMore) null else page.cursor,
            watermark = page.watermark,
            phase = if (page.hasMore) WorkProjectionPhase.BOOTSTRAPPING else WorkProjectionPhase.CURRENT,
            nextPageToken = page.nextPageToken,
            resetLedgerHint = null,
        ),
        subjects,
    )
}

private fun applyDelta(
    state: WorkProjection,
    page: WorkSyncPage,
    context: WorkSyncRequestContext,
): WorkProjection {
    val scope = WorkSyncScope(context.gatewayId, context.profileId)
    if (!sameScope(state, scope)) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.IDENTITY_CHANGED,
            "The delta belongs to a different gateway or profile; bootstrap first.",
        )
    }
    if (state.phase != WorkProjectionPhase.CURRENT && state.phase != WorkProjectionPhase.SYNCING) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.BOOTSTRAP_REQUIRED,
            "A delta cannot be applied before bootstrap completes.",
        )
    }
    if (state.ledgerId != page.ledgerId) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.LEDGER_CHANGED,
            "The Work ledger changed; discard the projection and bootstrap.",
        )
    }
    val stateCursor = state.cursor ?: throw WorkSyncApplyException(
        WorkSyncApplyErrorCode.BOOTSTRAP_REQUIRED,
        "The projection has no durable cursor.",
    )
    val requestedAfter = context.after ?: stateCursor
    if (requestedAfter < 0) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.CURSOR_INVALID,
            "The requested Work cursor is invalid.",
        )
    }
    if (requestedAfter != stateCursor) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.CURSOR_INVALID,
            "A stale or future Work response cannot advance this projection.",
        )
    }
    if (page.cursor < stateCursor || page.watermark < stateCursor) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.CURSOR_INVALID,
            "The Work page is behind the persisted cursor.",
        )
    }
    if (page.cursor > stateCursor && page.events.isEmpty()) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.CURSOR_INVALID,
            "A Work cursor cannot advance without the intervening events.",
        )
    }
    if (page.hasMore && page.cursor == stateCursor) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.CURSOR_INVALID,
            "A truncated Work page must advance beyond the persisted cursor.",
        )
    }

    val subjects = mutableSubjects(state)
    var expectedEventId = stateCursor + 1
    for (event in page.events) {
        // Replayed committed events cannot overwrite a newer subject.
        if (event.eventId <= stateCursor) continue
        if (event.eventId != expectedEventId) {
            throw WorkSyncApplyException(
                WorkSyncApplyErrorCode.CURSOR_INVALID,
                "The Work page skipped one or more events after the persisted cursor.",
            )
        }
        applyEvent(subjects, event)
        expectedEventId += 1
    }
    if (expectedEventId - 1 != page.cursor) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.CURSOR_INVALID,
            "The Work page cursor does not match its final contiguous event.",
        )
    }
    return finishProjection(
        WorkProjection(
            gatewayId = state.gatewayId,
            profileId = state.profileId,
            ledgerId = state.ledgerId,
            cursor = page.cursor,
            watermark = page.watermark,
            phase = if (page.hasMore) WorkProjectionPhase.SYNCING else WorkProjectionPhase.CURRENT,
            nextPageToken = null,
            resetLedgerHint = null,
        ),
        subjects,
    )
}

/** Apply one verified page atomically. */
fun applyWorkSyncPage(
    state: WorkProjection,
    page: WorkSyncPage,
    context: WorkSyncRequestContext,
): WorkProjection {
    val scope = checkedContext(context)
    if (page.workProfileId != scope.profileId) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.IDENTITY_CHANGED,
            "The Work page belongs to a different profile.",
        )
    }
    if (!page.actionable) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.PAGE_NON_ACTIONABLE,
            "This client does not understand the Work sync page mode.",
        )
    }
    return when (page.mode) {
        "bootstrap" -> applyBootstrap(state, page, context)
        "delta" -> applyDelta(state, page, context)
        else -> throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.PAGE_NON_ACTIONABLE,
            "This client does not understand the Work sync page mode.",
        )
    }
}

/** A cursor reset discards the old ledger immediately; bootstrap replaces it. */
fun applyWorkCursorReset(
    state: WorkProjection,
    reset: WorkCursorReset,
    scope: WorkSyncScope,
): WorkProjection {
    val checked = checkedScope(scope)
    if (!sameScope(state, checked)) {
        throw WorkSyncApplyException(
            WorkSyncApplyErrorCode.IDENTITY_CHANGED,
            "A cursor reset from a different gateway or profile was ignored.",
        )
    }
    return createWorkProjection(checked).copy(resetLedgerHint = reset.data.ledgerId)
}

File: apps/shared/src/work-sync.ts
import {
  type WorkAttention,
  type WorkCursorReset,
  type WorkEvent,
  type WorkEventSubject,
  type WorkJobSummary,
  type WorkSyncPage,
} from "./work-contract";

export interface WorkSyncScope {
  gateway_id: string;
  profile_id: string;
}

export type WorkProjectionPhase =
  | "empty"
  | "bootstrapping"
  | "syncing"
  | "current";

export interface WorkUnknownProjectionSubject {
  subject_id: string;
  subject_type: string;
  version: number;
  subject: WorkEventSubject;
}

/**
 * Serializable reference projection. The `(gateway, profile, ledger, cursor)`
 * tuple is state authority; live `work.changed` hints never mutate it.
 */
export interface WorkProjection {
  gateway_id: string;
  profile_id: string;
  ledger_id: string | null;
  cursor: number | null;
  watermark: number | null;
  phase: WorkProjectionPhase;
  next_page_token: string | null;
  reset_ledger_hint: string | null;
  jobs: Readonly<Record<string, WorkJobSummary>>;
  attention: Readonly<Record<string, WorkAttention>>;
  unknown_subjects: Readonly<Record<string, WorkUnknownProjectionSubject>>;
  /** Includes tombstone versions, preventing stale subjects from resurrecting. */
  subject_versions: Readonly<Record<string, number>>;
}

export type WorkSyncApplyErrorCode =
  | "identity_changed"
  | "bootstrap_sequence_invalid"
  | "bootstrap_required"
  | "cursor_invalid"
  | "ledger_changed"
  | "page_non_actionable";

export class WorkSyncApplyError extends Error {
  readonly code: WorkSyncApplyErrorCode;

  constructor(code: WorkSyncApplyErrorCode, message: string) {
    super(message);
    this.name = "WorkSyncApplyError";
    this.code = code;
  }
}

export interface WorkSyncRequestContext extends WorkSyncScope {
  /** The token sent to fetch this bootstrap page; null for page one. */
  page_token?: string | null;
  /** The cursor sent to fetch this delta page. */
  after?: number;
}

function nonempty(value: string, field: string): string {
  if (!value.trim()) throw new TypeError(`${field} must be non-empty.`);
  return value;
}

function assertScope(scope: WorkSyncScope): WorkSyncScope {
  return {
    gateway_id: nonempty(scope.gateway_id, "gateway_id"),
    profile_id: nonempty(scope.profile_id, "profile_id"),
  };
}

export function createWorkProjection(scope: WorkSyncScope): WorkProjection {
  const checked = assertScope(scope);
  return {
    ...checked,
    ledger_id: null,
    cursor: null,
    watermark: null,
    phase: "empty",
    next_page_token: null,
    reset_ledger_hint: null,
    jobs: {},
    attention: {},
    unknown_subjects: {},
    subject_versions: {},
  };
}

function sameScope(state: WorkProjection, scope: WorkSyncScope): boolean {
  return (
    state.gateway_id === scope.gateway_id &&
    state.profile_id === scope.profile_id
  );
}

function subjectKey(type: string, id: string): string {
  return `${type}:${id}`;
}

interface MutableProjectionSubjects {
  jobs: Record<string, WorkJobSummary>;
  attention: Record<string, WorkAttention>;
  unknown: Record<string, WorkUnknownProjectionSubject>;
  versions: Record<string, number>;
}

function mutableSubjects(state?: WorkProjection): MutableProjectionSubjects {
  return {
    jobs: state === undefined ? {} : { ...state.jobs },
    attention: state === undefined ? {} : { ...state.attention },
    unknown: state === undefined ? {} : { ...state.unknown_subjects },
    versions: state === undefined ? {} : { ...state.subject_versions },
  };
}

function applyJob(
  mutable: MutableProjectionSubjects,
  job: WorkJobSummary,
): void {
  const key = subjectKey("job", job.job_id);
  if ((mutable.versions[key] ?? 0) >= job.version) return;
  mutable.jobs[job.job_id] = job;
  delete mutable.unknown[key];
  mutable.versions[key] = job.version;
}

function applyAttention(
  mutable: MutableProjectionSubjects,
  attention: WorkAttention,
): void {
  const key = subjectKey("attention", attention.attention_id);
  if ((mutable.versions[key] ?? 0) >= attention.version) return;
  mutable.attention[attention.attention_id] = attention;
  delete mutable.unknown[key];
  mutable.versions[key] = attention.version;
}

function applyEvent(
  mutable: MutableProjectionSubjects,
  event: WorkEvent,
): void {
  const key = subjectKey(event.subject_type, event.subject_id);
  if ((mutable.versions[key] ?? 0) >= event.subject_version) return;

  if (event.tombstone) {
    if (event.subject_type === "job") delete mutable.jobs[event.subject_id];
    if (event.subject_type === "attention") {
      delete mutable.attention[event.subject_id];
    }
    delete mutable.unknown[key];
    mutable.versions[key] = event.subject_version;
    return;
  }

  const subject = event.subject;
  if (subject === null) {
    throw new WorkSyncApplyError(
      "page_non_actionable",
      "A non-tombstone work event has no subject.",
    );
  }
  if (subject.object_type === "job") {
    applyJob(mutable, subject);
    return;
  }
  if (subject.object_type === "attention") {
    applyAttention(mutable, subject);
    return;
  }
  mutable.unknown[key] = {
    subject_id: event.subject_id,
    subject_type: event.subject_type,
    version: event.subject_version,
    subject,
  };
  mutable.versions[key] = event.subject_version;
}

function finishProjection(
  base: Omit<
    WorkProjection,
    "jobs" | "attention" | "unknown_subjects" | "subject_versions"
  >,
  subjects: MutableProjectionSubjects,
): WorkProjection {
  return {
    ...base,
    jobs: subjects.jobs,
    attention: subjects.attention,
    unknown_subjects: subjects.unknown,
    subject_versions: subjects.versions,
  };
}

function applyBootstrap(
  state: WorkProjection,
  page: WorkSyncPage,
  context: WorkSyncRequestContext,
): WorkProjection {
  const requestedToken = context.page_token ?? null;
  const isFirstPage = requestedToken === null;
  let subjects: MutableProjectionSubjects;

  if (isFirstPage) {
    // A fresh bootstrap is always an explicit projection replacement. This is
    // the only page type allowed to change gateway/profile/ledger identity.
    subjects = mutableSubjects();
  } else {
    if (!sameScope(state, context)) {
      throw new WorkSyncApplyError(
        "identity_changed",
        "A bootstrap continuation belongs to a different gateway or profile.",
      );
    }
    if (
      state.phase !== "bootstrapping" ||
      state.next_page_token !== requestedToken
    ) {
      throw new WorkSyncApplyError(
        "bootstrap_sequence_invalid",
        "Bootstrap page token does not match the pending page.",
      );
    }
    if (state.ledger_id !== page.ledger_id) {
      throw new WorkSyncApplyError(
        "ledger_changed",
        "The Work ledger changed during bootstrap; restart at page one.",
      );
    }
    if (state.watermark !== page.watermark) {
      throw new WorkSyncApplyError(
        "cursor_invalid",
        "The fixed bootstrap watermark changed between pages.",
      );
    }
    subjects = mutableSubjects(state);
  }

  // Everything below mutates private copies. No cursor/identity is published
  // until every subject in the page has applied successfully.
  for (const job of page.jobs) applyJob(subjects, job);
  for (const attention of page.attention) applyAttention(subjects, attention);

  return finishProjection(
    {
      gateway_id: context.gateway_id,
      profile_id: context.profile_id,
      ledger_id: page.ledger_id,
      cursor: page.has_more ? null : page.cursor,
      watermark: page.watermark,
      phase: page.has_more ? "bootstrapping" : "current",
      next_page_token: page.next_page_token,
      reset_ledger_hint: null,
    },
    subjects,
  );
}

function applyDelta(
  state: WorkProjection,
  page: WorkSyncPage,
  context: WorkSyncRequestContext,
): WorkProjection {
  if (!sameScope(state, context)) {
    throw new WorkSyncApplyError(
      "identity_changed",
      "The delta belongs to a different gateway or profile; bootstrap first.",
    );
  }
  if (state.phase !== "current" && state.phase !== "syncing") {
    throw new WorkSyncApplyError(
      "bootstrap_required",
      "A delta cannot be applied before bootstrap completes.",
    );
  }
  if (state.ledger_id !== page.ledger_id) {
    throw new WorkSyncApplyError(
      "ledger_changed",
      "The Work ledger changed; discard the projection and bootstrap.",
    );
  }
  if (state.cursor === null) {
    throw new WorkSyncApplyError(
      "bootstrap_required",
      "The projection has no durable cursor.",
    );
  }
  const requestedAfter = context.after ?? state.cursor;
  if (!Number.isSafeInteger(requestedAfter) || requestedAfter < 0) {
    throw new WorkSyncApplyError(
      "cursor_invalid",
      "The requested Work cursor is invalid.",
    );
  }
  if (requestedAfter !== state.cursor) {
    throw new WorkSyncApplyError(
      "cursor_invalid",
      "A stale or future Work response cannot advance this projection.",
    );
  }
  if (page.cursor < state.cursor || page.watermark < state.cursor) {
    throw new WorkSyncApplyError(
      "cursor_invalid",
      "The Work page is behind the persisted cursor.",
    );
  }
  if (page.cursor > state.cursor && page.events.length === 0) {
    throw new WorkSyncApplyError(
      "cursor_invalid",
      "A Work cursor cannot advance without the intervening events.",
    );
  }
  if (page.has_more && page.cursor === state.cursor) {
    throw new WorkSyncApplyError(
      "cursor_invalid",
      "A truncated Work page must advance beyond the persisted cursor.",
    );
  }

  const subjects = mutableSubjects(state);
  let expectedEventId = state.cursor + 1;
  for (const event of page.events) {
    // The persisted monotonic cursor is also the event-id dedupe fence. This
    // makes replaying an already-committed response harmless.
    if (event.event_id <= state.cursor) continue;
    // A valid Work ledger allocates one global event ID per committed event,
    // and retention removes only a contiguous prefix. A gap therefore means
    // this response skipped durable state and must not advance the cursor.
    if (event.event_id !== expectedEventId) {
      throw new WorkSyncApplyError(
        "cursor_invalid",
        "The Work page skipped one or more events after the persisted cursor.",
      );
    }
    applyEvent(subjects, event);
    expectedEventId += 1;
  }
  if (expectedEventId - 1 !== page.cursor) {
    throw new WorkSyncApplyError(
      "cursor_invalid",
      "The Work page cursor does not match its final contiguous event.",
    );
  }

  // Commit the page projection first, then expose its cursor in the returned
  // immutable value. If any event throws, callers retain `state` unchanged.
  return finishProjection(
    {
      gateway_id: state.gateway_id,
      profile_id: state.profile_id,
      ledger_id: state.ledger_id,
      cursor: page.cursor,
      watermark: page.watermark,
      phase: page.has_more ? "syncing" : "current",
      next_page_token: null,
      reset_ledger_hint: null,
    },
    subjects,
  );
}

/**
 * Atomically apply one already-verified page. Unknown object enums remain in
 * the projection for display but their individual `actionable` flag is false.
 */
export function applyWorkSyncPage(
  state: WorkProjection,
  page: WorkSyncPage,
  context: WorkSyncRequestContext,
): WorkProjection {
  assertScope(context);
  if (page.work_profile_id !== context.profile_id) {
    throw new WorkSyncApplyError(
      "identity_changed",
      "The Work page belongs to a different profile.",
    );
  }
  if (!page.actionable) {
    throw new WorkSyncApplyError(
      "page_non_actionable",
      "This client does not understand the Work sync page mode.",
    );
  }
  if (page.mode === "bootstrap") {
    return applyBootstrap(state, page, context);
  }
  if (page.mode === "delta") return applyDelta(state, page, context);
  throw new WorkSyncApplyError(
    "page_non_actionable",
    "This client does not understand the Work sync page mode.",
  );
}

/**
 * Consume a typed cursor reset. The old projection is discarded immediately;
 * only a subsequent page-one bootstrap may establish the replacement ledger.
 */
export function applyWorkCursorReset(
  state: WorkProjection,
  reset: WorkCursorReset,
  scope: WorkSyncScope,
): WorkProjection {
  assertScope(scope);
  if (!sameScope(state, scope)) {
    throw new WorkSyncApplyError(
      "identity_changed",
      "A cursor reset from a different gateway or profile was ignored.",
    );
  }
  return {
    ...createWorkProjection(scope),
    reset_ledger_hint: reset.data.ledger_id,
  };
}
